### PR TITLE
Feat/pi coding agent sdk wrapper

### PR DESCRIPTION
**Pull request by lqiu03**
## Summary

- add TraceRoot-owned instrumentation for `@earendil-works/pi-coding-agent`, emitting AGENT → LLM → TOOL spans for apps that embed the pi SDK directly
- anchor the AGENT root span on each `AgentSession.prompt()` call, so a full turn — auto-retries, compaction, mid-stream `steer()`/`followUp()` — stays in one trace and re-parents under any active host span
- dual-write model/token usage as both OpenTelemetry `gen_ai.*` and OpenInference `llm.*`, matching `claude-agent-sdk` so pi spans render identically
- ship in-tree as a single `src/pi.ts` (sibling to `claude-agent-sdk.ts`), wired through `instrumentModules.piCodingAgent` — no standalone package, no private exporter pipeline
- add 63 regression cases across two `pi-*` test files (span lifecycle, event ordering, retry/dispose/sweep, capture flags, truncation)

This PR is pi-only src.

## Usage

```ts
import { TraceRoot } from '@traceroot-ai/traceroot';
import * as pi from '@earendil-works/pi-coding-agent';

// Bare module ref — capture defaults on:
TraceRoot.initialize({ instrumentModules: { piCodingAgent: pi } });

// Or opt out of content capture:
TraceRoot.initialize({
  instrumentModules: {
    piCodingAgent: { module: pi, config: { captureContent: false, captureToolIo: false } },
  },
});
```

Instrumentation patches `AgentSession.prototype`, so any session the host constructs is traced automatically — no per-session wiring. There is no `apiKey`/`baseUrl` here: pi traces through the provider `TraceRoot.initialize()` already registered.

## What gets traced

A `prompt()` call produces one trace, AGENT → LLM → TOOL (tools nest under the LLM turn that requested them):

| Span | `openinference.span.kind` | Key attributes |
|---|---|---|
| `AgentSession.prompt` | `AGENT` (root) | `session.id`, `input.value`/`output.value` (when `captureContent`), `traceroot.pi.retry_count` |
| `<model>` | `LLM` | `gen_ai.request.model` + `llm.model_name`, `gen_ai.usage.input_tokens`/`output_tokens` + `llm.token_count.prompt`/`completion`/`total` (incl. cache read/write), `output.value` |
| `<tool>` | `TOOL` | `gen_ai.tool.name`, `gen_ai.tool.call.id`, `input.value`/`output.value` (when `captureToolIo`) |

- **One trace per turn.** Root opens on `prompt()`, closes when its promise settles; retries/compaction/follow-ups stay under it. `traceroot.pi.retry_count` records auto-retries.
- **Error scoping.** A failed tool marks only its TOOL span ERROR; a failed LLM turn marks only the LLM span — parents stay OK.

## Design

- **In-tree, single file**, mirroring `claude-agent-sdk.ts` in structure and attribute shape. `instrumentModules.piCodingAgent` accepts a bare module ref or the `{ module, config }` wrapper.
- **Prototype patching.** Patches `AgentSession.prototype.prompt/steer/followUp/dispose` (the prototype, not the frozen ESM namespace). Install is fail-loud on a bad SDK shape, idempotent on re-entry, and rolls back all patches on partial failure. `steer()`/`followUp()` are patched only so `subscribe()` runs once — either can be a session's first call.
- **Tracer resolved at use.** The tracer is resolved at each span-open rather than captured at wrap time, so pi spans always route to whichever provider is registered at that moment. Worth flagging that this alone does not survive a `shutdown()`/`initialize()` cycle on `main` today — `shutdown()` never releases the OTel global slots, so the second `initialize()`'s `register()` silently loses and the global keeps delegating to the old, already-shut-down provider. pi re-resolves correctly; there is just nothing new to resolve to. That is a `shutdown()` bug, not a pi bug, and is deliberately out of scope here; pi needs no further change when it is fixed.
- **Lifecycle sweep.** A per-session state machine force-closes dangling spans from a crashed attempt (`traceroot.pi.force_closed`) on `agent_start`, `turn_end`, `agent_end`, mid-run rejection, and `dispose()`; a session reused after `dispose()` re-subscribes on its next `prompt()`.
- **Payload safety.** Tool I/O is JSON-serialized with a 32 KB size cap and a 2,048-entry breadth cap, truncated with a plain slice; tool span names are command/basename only, never full paths.

## Files changed

New: `src/pi.ts`, `tests/pi.test.ts`, `tests/pi-lifecycle-edge-cases.test.ts`, `tests/pi-test-helpers.ts`.

Modified: `src/instrumentation.ts` (one import + one call, symmetric with claude/openai), `src/types.ts` (`PiCodingAgentInstrumentation`, `instrumentModules.piCodingAgent`), `src/index.ts` (type re-exports), `package.json` (two devDependencies).

No shared configuration or existing integration is touched: `tsconfig.json`, `src/constants.ts`, `src/claude-agent-sdk.ts` and `src/traceroot.ts` all stay at upstream.

On the lockfile: `pnpm-lock.yaml` shows a large diff that is not specific to this PR. `@arizeai/openinference-instrumentation-bedrock` peer-depends on `@aws-sdk/client-bedrock-runtime: ^3.0.0`, so pnpm re-resolves that caret — and the AWS SDK's transitive tree with it — on any `package.json` edit. Regenerating the lockfile on `main` with an unrelated one-line manifest change produces the same churn. CI installs with `--frozen-lockfile`, which checks lockfile-manifest consistency rather than resolution freshness, so this is safe to take as-is; happy to hand-prune it to just the two new devDependencies if you'd rather keep the diff minimal.

## Testing

- `npm run build` / typecheck — clean
- unit: 63 `it()` cases across the two `pi-*` files (span shape, context parenting, retry/dispose/sweep, capture-flag boundaries, truncation); full package suite 227/227 green
- retry behavior (`retry_count`, ERROR→OK LLM child on retry) is covered by a mocked stream — not reliably reproducible on a live happy-path run
- **live e2e** against Traceroot cloud with this branch's build: real `AgentSession` running the pi incident-response example (traceroot#1547) on `gpt-5.4`, exported HTTP 200

## End-to-end tests
Screenshot 2026-07-24 at 7 06 18 PM
Test 1 - (https://github.com/traceroot-ai/traceroot/pull/1547)

Screenshot 2026-07-24 at 7 06 53 PM
Test 2

### CODE REVIEW

**Comment by lqiu03** [comment]
@cubic-dev-ai quickly review please

File: packages/pi/tests/dangling-span-sweep-deduplication.test.ts [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+  // misbehaving span can never abort the rest of the sweep.
+  const helperDefinitions = SOURCE.match(/function safeCloseDanglingSpan\(/g) ?? [];
+  assert.equal(
+    helperDefinitions.length,
+    1,
+    'expected exactly one safeCloseDanglingSpan() helper definition — found ' +
+      `${helperDefinitions.length}.`,
+  );
+  const bareCloseInSweep =
+    SOURCE.match(/state\.toolSpans\.clear\(\);\s*closeDanglingSpan\(/g) ?? [];
```
[the remark this answers is not in the document]
Comment (lqiu03) [reply packages/pi/tests/dangling-span-sweep-deduplication.test.ts]: Claimed the test regex closeDanglingSpan\( false-matches afeCloseDanglingSpan(. It doesn't — the match is case-sensitive and he substring inside the camelCase name is CloseDanglingSpan (capital C), not closeDanglingSpan. Ran the actual regex against the real ile myself: zero matches

File: packages/pi/src/instrumentation.ts [hunk 2]
```diff
@@ -0,0 +1,758 @@
...
+    // this proactive branch and skip queuing — losing that real run's
+    // input.value (falling back to agent_start's empty-queue heuristic)
+    // rather than corrupting a LATER call's attribution. That tradeoff —
+    // graceful degradation (a missing/stale input.value) over the
+    // cross-call corruption this queue exists to prevent — is preferred
+    // given the SDK exposes no way to precisely predict a specific hook's
+    // decision without invoking it. This (and the enumeration of early-return
+    // paths above) must be kept in sync with the SDK and may miss future
+    // early-return paths a later SDK version adds.
+    const mayBeHandledByInputHook = this.hasExtensionHandlers?.('input') === true;
```
[the remark this answers is not in the document]
Comment (lqiu03) [reply packages/pi/src/instrumentation.ts]: I'd push back and accept that this is a tradeoff - in fact, your fix @cubic-dev-ai would introduce a bug regression. preflightResult?.(true) fires identically on both intercepted-command path and the agent-start path

File: packages/pi/tests/shared-mode-no-apikey.test.ts [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+import { test } from 'node:test';
+import { trace } from '@opentelemetry/api';
+import { NodeTracerProvider } from '@opentelemetry/sdk-trace-node';
+import { InMemorySpanExporter, SimpleSpanProcessor } from '@opentelemetry/sdk-trace-base';
+import { instrumentPiCodingAgent } from '../src/instrumentation';
+import { assistantMessage, attrs, makeFakeSessionClass } from './test-helpers';
+
+// The whole point of this regression is a genuinely-absent apiKey, so pin
+// TRACEROOT_API_KEY to unset for the duration regardless of the ambient env.
+function withoutApiKeyEnv(fn: () => Promise<void> | void): Promise<void> | void {
```
[the remark this answers is not in the document]
Comment (lqiu03) [reply packages/pi/tests/shared-mode-no-apikey.test.ts]: Defer; this is a test script and the mechanism is never touched in prod

File: packages/traceroot/tests/pi-coding-agent-integration.test.ts [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+
+afterEach(() => {
+  _resetForTesting();
+});
+
+test('the real @traceroot-ai/pi wired through TraceRoot.initialize exports enriched spans into TraceRoot own pipeline', async () => {
+  // Unset so the apiKey can only reach pi via initialize()'s config plumbing,
+  // never the TRACEROOT_API_KEY env fallback — this is the P0 that made pi
+  // silently no-op before the fix.
+  delete process.env.TRACEROOT_API_KEY;
```
[the remark this answers is not in the document]
Comment (lqiu03) [reply packages/traceroot/tests/pi-coding-agent-integration.test.ts]: Defer; this is a test script, plus delete process.env.TRACEROOT_API_KEY isn't restored in afterEach, not applicable

File: packages/pi/src/instrumentation.ts [hunk 5]
```diff
@@ -0,0 +1,964 @@
...
+// queue holds nothing but stale entries (or is empty), so agent_start falls
+// back to its usual state.pendingInputText path instead of trusting stale
+// text. Warns once per dequeue that discarded anything, matching this file's
+// other console.warn conventions, so a real host-side "prompt never started"
+// anomaly is visible rather than silently swallowed.
+function dequeueFreshQueuedPrompt(queue: QueuedPrompt[]): QueuedPrompt | undefined {
+  const now = Date.now();
+  let discarded = 0;
+  let entry = queue.shift();
+  while (entry && now - entry.enqueuedAt > MAX_QUEUED_PROMPT_AGE_MS) {
```
[the remark this answers is not in the document]
Comment (lqiu03) [reply packages/pi/src/instrumentation.ts]: Defer; need a proper impl mechanism for even based correlation instead of timeout

File: packages/pi/src/provider.ts [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+// Going through the global `trace` facade (rather than duck-typing
+// getDelegate()) is also inherently robust to a dual-copy @opentelemetry/api
+// install: the global provider registration lives in globalThis-based storage
+// keyed by Symbol.for(`opentelemetry.js.api.<major>`) (see @opentelemetry/api's
+// own internal/global-utils.ts) and is shared across copies of the same major
+// version, so the probe always resolves to the genuinely-registered provider —
+// even though a nominal instanceof check would be defeated by the cross-copy
+// class-identity mismatch.
+function hasRealGlobalProvider(): boolean {
+  const probe = trace.getTracer(SDK_NAME, SDK_VERSION).startSpan('traceroot-pi.provider-probe');
```
[the remark this answers is not in the document]
Comment (lqiu03) [reply packages/pi/src/provider.ts]: A tradeoff I'm willing to accept (reverting it would result in a spam trace into user's trace list); defer

**Comment by lqiu03** [comment]
**New E2e tests:**

Screenshot 2026-07-14 at 10 13 06 PM
Run 1: no-tool use, pure Q&A

Screenshot 2026-07-14 at 10 15 41 PM
Run 2: Search for a file path that doesn't exist

Screenshot 2026-07-14 at 10 18 45 PM
Run 3: a two-turn agent

Screenshot 2026-07-14 at 10 19 13 PM
Run 4: a more complicated agent with multiple steps, reasoning, tool call, and planted errors to mimick a potential blocker an agent might face in a real workflow~

**Comment by lqiu03** [comment]
**New Updates:** 

TraceRoot can now record what a Pi agent does while it runs. As Pi works, it emits events — a run starts, the model gets called, a tool runs, an error hits — and TraceRoot hooks each event and turns it into a timestamped span holding the prompt, the reply, each tool’s inputs and outputs, how long it took, and how many tokens were spent. The spans from one run are stitched into a single trace and shipped to your dashboard, where you can walk the run step by step. Like Traceroot Claude SDK, Pi feeds into that same export-and-render path rather than a parallel one, so nothing new was needed on the receiving end. You switch it on by registering Pi at initialization, and Pi can also run on its own, tracing Pi sessions with none of the rest of TraceRoot installed.

**Summary of what Pi Traceroot SDK records for every Pi agent session**
- Every exchange with the AI model: model, tokens, cost, input/output.
- Every tool the agent uses (file read/write/edit, search, running commands) — the exact call, the result, and whether it failed.
- The shape of the session: which tool call belongs to which AI response, in order, correctly nested.
- A failure in one step stays scoped to that step — it does not mark the whole session as broken.

File: packages/traceroot/tests/initialize-lifecycle-and-wiring-dispatch.test.ts [hunk 7]
```diff
@@ -0,0 +1,282 @@
...
+    );
+  });
+});
+
+describe('beforeExit listener hygiene across init/shutdown cycles', () => {
+  it('does not accumulate a beforeExit listener per initialize() that shutdown() never removes', async () => {
+    delete process.env.TRACEROOT_API_KEY;
+    const before = process.listenerCount('beforeExit');
+
+    // Three full, clean lifecycles. Each successful initialize() registers a
```
[the remark this answers is not in the document]
Comment (lqiu03) [reply packages/traceroot/tests/initialize-lifecycle-and-wiring-dispatch.test.ts]: Defer; this is a test script, not applicable to prod code

**Comment by lqiu03** [comment]
**More end-to-end tests:**

Screenshot 2026-07-15 at 5 59 12 PM
Screenshot 2026-07-15 at 5 55 12 PM
Run 1: Running concurrent sessions to stress-test

Screenshot 2026-07-15 at 5 56 25 PM
Run 2; multiple turn agent for long-running sessions

Screenshot 2026-07-15 at 5 57 07 PM
Run 3: pi agent ran into errors, and is writing code to fix issues

Screenshot 2026-07-15 at 5 57 49 PM
Run 4: canceling the agent turn as soon as the agent is started

File: packages/traceroot/src/pi/config.ts [hunk 8]
```diff
@@ -0,0 +1,32 @@
+/**
+ * Configuration resolution for the Pi coding agent instrumentation.
+ *
+ * Mirrors the config-resolution contract shared by every traceroot-ts
+ * integration: explicit config wins, then env vars, then a built-in default.
```
[the remark this answers is not in the document]
Comment (lqiu03) [reply packages/traceroot/src/pi/config.ts]: I would defer but ack
